src/foam/nanos/auth/ServiceProviderAwareSupport.js

```diff
@@ -0,0 +1,169 @@
+                                put(NotificationSetting.class.getName(), new PropertyInfo[] { NotificationSetting.OWNER });
+                              }}
+                            )
+                            .build();
+
+`,
+
+  // TODO: ServiceProviderAwareMLang(property) performs EQ on getSpid on both objs.
+  // and add to generation where 'findFoo' is available to avoid reflection.,
+
+  javaImports: [
+    'foam.core.PropertyInfo',
+    'foam.core.X',
+    'foam.nanos.logger.Logger',
+    'java.util.HashMap',
+    'java.util.Map'
+  ],
+
+  properties: [
+    {
+      name: 'findMethods',
+      class: 'Map',
+      visibility: 'HIDDEN',
+      javaFactory: 'return new HashMap();'
+    }
+  ],
+
+  methods: [
+    {
+      documentation: `Using Relationship findFoo(x), traverse relationships,
+returning true if the context users spid matches the current object.`,
+      name: 'match',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'properties',
+          type: 'Map'
+        },
+        {
+          name: 'obj',
+          type: 'Object'
+        }
+      ],
+      type: 'Boolean',
+      javaCode: `
+      User user = (User) x.get("user");
+      if ( user == null ) {
+        // TODO/REVIEW: occurs during login. See AuthenticationApiTest
+        return true;
+      }
+
+      if ( obj != null &&
+           obj instanceof ServiceProviderAware ) {
+        return ((ServiceProviderAware) obj).getSpid().equals(user.getSpid());
+      }
+
+      Object result = obj;
+      while ( result != null ) {
+        PropertyInfo pInfos[] = (PropertyInfo[]) properties.get(result.getClass().getName());
+        if ( pInfos == null ) {
+          return false;
+        }
+        for ( int i = 0; i < pInfos.length; i++) {
+          foam.core.PropertyInfo pInfo = pInfos[i];
+          try {
+            java.lang.reflect.Method method = getFindMethod(x, pInfo.getName(), result);
+            result = invokeMethod(x, method, result);
+            if ( result != null &&
+                 result instanceof ServiceProviderAware ) {
+              return ((ServiceProviderAware) result).getSpid().equals(user.getSpid());
+            }
+          } catch (Throwable e) {
+            return false;
+          }
+        }
+      }
+      return false;
+      `
+    },
+    {
+      name: 'getFindMethod',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'name',
+          type: 'String'
+        },
+        {
+          name: 'obj',
+          type: 'Object'
+        }
+      ],
+      type: 'java.lang.reflect.Method',
+      javaThrows: ['Exception'],
+      javaCode: `
+    java.lang.reflect.Method method = (java.lang.reflect.Method) getFindMethods().get(name);
+    if ( method == null ) {
+      String methodName = "find" + name.substring(0,1).toUpperCase() + name.substring(1);
```

Comment:
we would also need to handle capitalization for multi-word property name such as userId - https://github.com/foam-framework/foam2/blob/master/src/foam/nanos/notification/Notification.js#L79.

Reply:
Yep, the real solution is to refine PropertyInfo with the Reference methods, but have no idea how to yet.

Reply:
multi-word should work.  If the name is userId,  find becomes findUserId, so just the first letter needs to be uppercased.

Reply:
that's right, it should work fine.